File: src/algorithms/sorting/comparison/cycle-sort/sources/cycle-sort.ts
// Cycle Sort — for each element, count elements smaller than it to find its correct position;
// place it there. Minimizes the number of writes to the array.

function cycleSort(inputArray: number[]): number[] {
  // @step:initialize
  const sortedArray = [...inputArray]; // @step:initialize
  const arrayLength = sortedArray.length; // @step:initialize

  for (let cycleStart = 0; cycleStart < arrayLength - 1; cycleStart++) {
    // @step:count-position
    let currentValue = sortedArray[cycleStart]!; // @step:count-position

    // Find the correct position for currentValue
    let correctPosition = cycleStart; // @step:count-position
    for (let scanIndex = cycleStart + 1; scanIndex < arrayLength; scanIndex++) {
      // @step:compare
      if (sortedArray[scanIndex]! < currentValue) {
        // @step:compare
        correctPosition++; // @step:count-position
      }
    }

    // If the item is already in the correct position, skip this cycle
    if (correctPosition === cycleStart) {
      continue; // @step:count-position
    }

    // Skip over duplicates to find the unique insertion point
    while (currentValue === sortedArray[correctPosition]!) {
      // @step:count-position
      correctPosition++; // @step:count-position
    }

    // Place currentValue at its correct position
    const displacedValue = sortedArray[correctPosition]!; // @step:swap
    sortedArray[correctPosition] = currentValue; // @step:swap
    currentValue = displacedValue; // @step:swap

    // Rotate the rest of the cycle
    while (correctPosition !== cycleStart) {
      // @step:count-position
      correctPosition = cycleStart; // @step:count-position

      for (let scanIndex = cycleStart + 1; scanIndex < arrayLength; scanIndex++) {
        // @step:compare
        if (sortedArray[scanIndex]! < currentValue) {
          // @step:compare
          correctPosition++; // @step:count-position
        }
      }

      while (currentValue === sortedArray[correctPosition]!) {
        // @step:count-position
        correctPosition++; // @step:count-position
      }

      if (currentValue !== sortedArray[correctPosition]!) {
        // @step:swap
        const nextDisplacedValue = sortedArray[correctPosition]!; // @step:swap
        sortedArray[correctPosition] = currentValue; // @step:swap
        currentValue = nextDisplacedValue; // @step:swap
      }
    }

    // @step:mark-sorted
  }

  // @step:mark-sorted
  return sortedArray; // @step:complete
}

File: src/algorithms/sorting/comparison/cycle-sort/sources/cycle-sort.rs
// Cycle Sort — for each element, count elements smaller than it to find its correct position;
// place it there. Minimizes the number of writes to the array.
fn cycle_sort(input_array: &[i64]) -> Vec<i64> {
    // @step:initialize
    let mut sorted_array = input_array.to_vec(); // @step:initialize
    let array_length = sorted_array.len(); // @step:initialize

    for cycle_start in 0..array_length.saturating_sub(1) {
        // @step:count-position
        let mut current_value = sorted_array[cycle_start]; // @step:count-position

        // Find the correct position for current_value
        let mut correct_position = cycle_start; // @step:count-position
        for scan_index in (cycle_start + 1)..array_length {
            // @step:compare
            if sorted_array[scan_index] < current_value {
                // @step:compare
                correct_position += 1; // @step:count-position
            }
        }

        // If the item is already in the correct position, skip this cycle
        if correct_position == cycle_start {
            continue; // @step:count-position
        }

        // Skip over duplicates to find the unique insertion point
        while current_value == sorted_array[correct_position] {
            // @step:count-position
            correct_position += 1; // @step:count-position
        }

        // Place current_value at its correct position
        let displaced_value = sorted_array[correct_position]; // @step:swap
        sorted_array[correct_position] = current_value; // @step:swap
        current_value = displaced_value; // @step:swap

        // Rotate the rest of the cycle
        while correct_position != cycle_start {
            // @step:count-position
            correct_position = cycle_start; // @step:count-position

            for scan_index in (cycle_start + 1)..array_length {
                // @step:compare
                if sorted_array[scan_index] < current_value {
                    // @step:compare
                    correct_position += 1; // @step:count-position
                }
            }

            while current_value == sorted_array[correct_position] {
                // @step:count-position
                correct_position += 1; // @step:count-position
            }

            if current_value != sorted_array[correct_position] {
                // @step:swap
                let next_displaced_value = sorted_array[correct_position]; // @step:swap
                sorted_array[correct_position] = current_value; // @step:swap
                current_value = next_displaced_value; // @step:swap
            }
        }

        // @step:mark-sorted
    }

    // @step:mark-sorted
    sorted_array // @step:complete
}
